pytransform3d/rotations/_conversions.py

```diff
@@ -1383,6 +1383,122 @@ def extrinsic_euler_zxy_from_active_matrix(R, strict_check=True):
         R, unity, unitx, unitz, False, strict_check)[::-1]
 
 
+def euler_from_quaternion(q, i, j, k, extrinsic):
+    """General method to extract any Euler angles from quaternions.
+
+    Parameters
+    ----------
+    q : array-like, shape (4,)
+        Unit quaternion to represent rotation: (w, x, y, z)
+
+    i : int
+        Axis of the first rotation angle. A number between 0 and 2.
+
+    j : int
+        Axis of the second rotation angle. A number between 0 and 2.
+
+    k : int
+        Axis of the third rotation angle. A number between 0 and 2.
+
+    extrinsic : bool
+        Do we use extrinsic transformations? Intrinsic otherwise.
+
+    Returns
+    -------
+    euler_angles : array, shape (3,)
+        Extracted rotation angles in radians about the axes i, j, k in this
+        order. The first and last angle are normalized to [-pi, pi]. The middle
+        angle is normalized to either [0, pi] (proper Euler angles) or
+        [-pi/2, pi/2] (Cardan / Tait-Bryan angles).
+
+    References
+    ----------
+    Bernardes, Evandro; Viollet, Stephane: Quaternion to Euler angles
+    conversion: A direct, general and computationally efficient method,
+    https://doi.org/10.1371/journal.pone.0276302
+    """
+    q = check_quaternion(q)
+
+    i += 1
+    j += 1
+    k += 1
+
+    # The original algorithm assumes extrinsic convention. Hence, we swap
+    # the order of axes for intrinsic rotation.
+    if not extrinsic:
+        i, k = k, i
+
+    # Proper Euler angles rotate about the same axis in the first and last
+    # rotation. If this is not the case, they are called Cardan or Tait-Bryan
+    # angles and have to be handled differently.
+    proper_euler = i == k
+    if proper_euler:
+        k = 6 - i - j
+
+    sign = (i - j) * (j - k) * (k - i) // 2
+    a = q[0]
+    b = q[i]
+    c = q[j]
+    d = q[k] * sign
+
+    if not proper_euler:
+        a, b, c, d = a - c, b + d, c + a, d - b
+
+    # Equation 35
+    angle_j = math.acos(
+        2.0 * (a * a + b * b) / (a * a + b * b + c * c + d * d) - 1.0)
+    # or Equation 34; TODO which one should be used and why?
+    #angle_j = 2.0 * math.atan2(math.sqrt(c * c + d * d),
```

Comment:
This is a good question. 
I had opted for Equation 35 but ended up changing to Equation 34 during the Scipy implementation.
Apparently, `acos` and `asin` implementations might have slight numerical problems near `zero`, making it preferable to use `atan2` whenever possible.
I would implement this as:
```
2 * math.atan2(math.hypot(c, d), math.hypot(a, b))
```